### PR TITLE
Add a header menu to switch between edit and select tool

### LINKED ISSUE #17088: Visually indicate navigation mode, and make mouse accessible
*This ticket has been updated in light of changes recently shipped. You can see the old version in the edit history.*

**Challenge**: Blocks can nest to any depth, but selecting the right level is fiddly. E.g. it's easy to click a Paragraph nested inside a Columns block to edit it, but it's fiddly and unintuitive to select the parent _column_ block to change its width.

## "Navigation Mode"

WordPress 5.3 has just been released with a new accessibility feature called "Navigation Mode":

- Press Escape to enter this mode.
- You can now use the arrow-keys or tab, to quickly navigate blocks.
- Press Enter or click a block to edit it. 

This improves keyboard accessibility by enabling you to tab through 10 blocks in 10 tab presses, as opposed to having to tab through all the block UI along the way. Another benefit is that in this mode, blocks are effectively _selected_, so you can press Delete to remove them. E.g. to delete a Paragraph, you can now press Escape to select it, press Delete, and press Enter to go back to editing.

**However there are some unaddressed challenges with this mode:**

- There is no visual indication of which mode you're in, other than the different block selection style.
- There is no mouse affordance to pick between modes.
## Suggested Solution

By reframing "navigation mode" as tools, we might be able to address the discoverability issues, and provide an interface for selecting nested blocks in a parent-first fashion for mouse users. 

**1. Visually indicate the mode**

Edit Tools Dropdown

By creating a visual affordance for showing what tool is selected, we are effectively acknowledging and visualizing the change that happens when you press Escape. 

Select Tools Dropdown

Additionally it provides a spot for reminding you of the shortcut keys, explaining what's going on, and provide a mouse interface for switching between the two.

For the majority of the time, users will simply see this interface:

Edit Tool

But now there will be a clear non-default affordance to choose to enter "navigation mode"/choose the "Select" tool.

*As an additional enhancement, it would likely be helpful to remove the block hover style from this default interface. "Hunting pixels" to select the parent block, as opposed to using the breadcrumb bar or the Select tool, is not an experience worth indicating with a hover style, that will "flicker" in narrow contexts.*

**2. Add parent-first selection**

When you select the **Select tool**, or press Escape, blocks are selected "parent-first", i.e. click any block to select the nesting container, then the first immediate child, and so on. E.g. click a group block to select it, then you can click child blocks after, to select them.

_This parent-first selection model you may recognize from Figma, or Sketch, or Adobe Illustrator, where grouped items are selected similarly. Only there, you double-click to enter a group. It has also been tested as the default interface in the Gutenberg plugin, under the name "clickthrough"._

Select Tool

- The style for selected and hovered blocks is inspired by Figma, with a 2px outline when hovered, and a 1px outline and underlined text for selected. It also absorbs, entirely, the block breadcrumb from the default interface.
- For blocks with nesting, you click through each layer. Columns > Column > Paragraph. 

To exit the selection tool you can either: 

- Press Enter to edit the selected block.
- Choose the Edit tool from the toolbar.
- Click twice on any block, at its deepest nesting level:
  - I.e. click a Paragraph to select it, click it again to edit it.
  - Click a group block to select it. Click a paragraph inside to select it. Click the Paragraph again to edit it. 

Full flow:

Edit   Selection Modes i2

## Finding The Balance

Feedback suggests that we should be mindful about introducing new modes. The approach outlined in this ticket has been updated to try and address that by _not introducing a mode_, but rather enhancing the existing mode "navigation mode". 

The goal is to both enhance that mode by adding a clear visual tool indicator, and by improving affordances for selecting those tools. At the same time, hopefully we can use this new context to provide advanced block selection tools, only when you need them.

### LINKED ISSUE #17847: Accessibility: tab navigation changed since WordPress 5.2
When using WordPress' current stable release, you can rely on the `Tab` key to quickly start composing:
1. Focus is set to the title as you open the editor.
2. Once you've added a title, you can hit `Tab` to quickly get to the main content area, where you'll be able to add blocks (by hitting `/` for example).

This changed with the current version of Gutenberg.

1. Go to Posts > Add New
2. Enter a post title
3. Hit Tab.
4. Start typing. It turns out you are not in the main content area.

**Desktop (please complete the following information):**
 - OS: macOS Catalina
 - Browser: Chrome
 - Version: 77.0.3865.90

### LINKED ISSUE #18583: Accessibility regression: Focus does not land in the paragraph for a new block when pressing Enter
**Describe the bug**
When starting a post and after entering the title, or after writing a paragraph, pressing Enter no longer puts focus in the paragraph of the newly added block. Instead, focus goes to the "Block" button above it. Windows screen readers such as NVDA then switch out of focus mode into virtual, browse, or reading mode, breaking the typing flow. The user has to down arrow and press Enter to put focus back into the paragraph to continue typing. This is very disruptive when thoughts are flowing and one expects to just be able to continue typing when pressing Enter for a new paragraph.

**To reproduce**
1. Go to your blog and start a new post. This also reproduces on wordpress.com, doesn't need to be self-hosted.
2. Type your title.
3. Press Enter.
4. Focus lands on the button that is either labeled Block or Paragraph above the actual paragraph.

**Expected behavior**
Focus lands in the paragraph of the newly added block so I can continue typing, or type a / to choose a new block type.

**Screenshots**
I am blind, so screenshots wouldn't do much good.

**Desktop (please complete the following information):**
 - OS: Windows 10.
 - Browser Firefox
 - Version 70, 71 beta, or 72 nightly.
 - Screen reader: NVDA 2019.2.1.

**Additional context**
From a chat in the WordPress accessibility channel on Slack, @aduth suspects this to be a regression from #16500. Similar to #17776.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #17088  
closes #17847
closes #18583 

First attempt at making the select/edit tool more visible in the UI.

This PR does:

 - Adds a dropdown into the header to switch between select and edit tools.
 - Remove the explicit switch to the edit tool as soon as you click the canvas.
 - Enables the edit tool by default.
 - Enables clickthrough in select tool.

Capture d’écran 2019-11-20 à 10 20 11 AM

### CODE REVIEW

**Comment by gziolo** [comment]
Can you include some visuals to make it easier to share feedback, please? :)

**Comment by gziolo** [comment]
I couldn't resist to add it:

Screen Shot 2019-11-20 at 10 18 33

I have some feedback to share:
- the label in the tooltip should follow the pattern and be shortened to `Navigation tool`
- it doesn't behave like a menu, you can compare with the More Menu:
    - you can't use arrow-down key to open the menu
    - you can use arrow-up and arrow-down to switch between menu items
- there is also one usability challenge, the shortcut for `Select` mode is `esc`, however when focus is in the dropdown menu – `esc` will close the menu, it might be confusing as some user might assume that this will change the mode, I don't have a solution for it

**Comment by youknowriad** [comment]
> it doesn't behave like a menu, you can compare with the More Menu:
 +
I actually used `DropdownMenu` first, but there's an inconsistency in styling between `Dropdown + IcconButton` and `DropdownMenu` that let me to fallback to `Dropdown` In addition to that, AFAIK we can add "help text" like the sentence at the end in `DropdownMenu` at the moment.

**Comment by gziolo** [comment]
Some more visuals for reference:
Screen Shot 2019-11-20 at 10 26 02
Screen Shot 2019-11-20 at 10 27 08
Screen Shot 2019-11-20 at 10 26 45

I don't think we are consistent at all so the part about the label would be nice to clarify in general, in addition we should decide whether we should include the heading in  🤷‍♂

**Comment by gziolo** [comment]
It seems to work but would require some CSS tweaks:
```diff
diff --git a/packages/edit-post/src/components/header/more-menu/index.js b/packages/edit-post/src/components/header/more-menu/index.js
index 1f781b513..f03b8aa69 100644
--- a/packages/edit-post/src/components/header/more-menu/index.js
+++ b/packages/edit-post/src/components/header/more-menu/index.js
@@ -36,6 +36,9 @@ const MoreMenu = () => (
                                <MenuGroup>
                                        <OptionsMenuItem />
                                </MenuGroup>
+                               <div className="block-editor-navigation-mode-selector__help">
+                                       { __( 'Tools offer different block interactions to optimize block selection & editing tasks' ) }
+                               </div>
                        </>
                ) }
        </DropdownMenu>
```
It also would work with `div` inside `MenuGroup`.
 
Screen Shot 2019-11-20 at 10 40 04

It now makes me wonder whether we could extend the API of `MenuGroup` to support such help text.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (gziolo) [inline packages/block-editor/src/components/inner-blocks/index.js]: Now the question is how we should approach mobile in this context. Should we still offer the option to switch between tools or should we hide them and provide the unique configuration tailored to touch screens.

Comment (jasmussen) [reply packages/block-editor/src/components/inner-blocks/index.js]: I have on my ToDo list to thoroughly review this, because I love this PR. But this just a quick note:

> Now the question is how we should approach mobile in this context. Should we still offer the option to switch between tools or should we hide them and provide the unique configuration tailored to touch screens.

It could be worth exploring this at some point. But there are some other aspects that probably should precede this. Right now the mobile interface is barely usable, and we need to "free up some space" to enable this to happen. ios 13 may enable a top toolbar, which could help, and breadcrumbs have not yet been brought to that interface — I know @iamthomasbishop has thought about how to do that. Depending on how that works, it might allow us to enforce _breadcrumb only_ navigation. Until then, I think we should continue to enforce clickthrough only on mobile.

Comment (youknowriad) [reply packages/block-editor/src/components/inner-blocks/index.js]: > ios 13 may enable a top toolbar

This is on my plate too, but it's blocked by #18044 to make the change simple.

Comment (jasmussen) [reply packages/block-editor/src/components/inner-blocks/index.js]: Yep, I forgot to suggest that I think that's a totally separate issue to explore and this PR should not be blocked by it.

**Comment by youknowriad** [comment]
Updated this based on the feedback, it should be better now.

**Comment by jasmussen** [comment]
Finally looking at this. Even as just a visualization of Navigation Mode, this feels like a step in the right direction. I would encourage anyone to test this PR out in practice.

Can you try this icon for the select tool instead?

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M6.5 1v21.5l6-6.5H21L6.5 1zm5.1 13l-3.1 3.4V5.9l7.8 8.1h-4.7z" /></svg>
```

Can you change the tooltip to just say "Tools"?

This also, in practice, feels like a great place to re-introduce the parent-first/clickthrough interface. I think I might either push some hover effect changes directly to this branch, or if you prefer, in a followup PR, to better help ambiguate the two modes and clarify the clickthrough effect better.

One thing worth doing, though — it's important that it's easy to return to "edit mode" trivially, even when using the mouse. Right now the only way to do that is to press Enter on a selected block, or selecting the edit tool. Can we add so this is also done when you click a second time on a selected block? I.e. this flow should work:

1. Press Escape to enter navigation/selection mode
2. Click a block to select it
3. **If the block you click is at the deepest nesting level**, clicking it again should exit back to edit mode. That is, if you click a Paragraph twice inside a Cover, you go back to editing it.

**Comment by youknowriad** [comment]
> I think I might either push some hover effect changes directly to this branch, or if you prefer, in a followup PR

I'm fine with both.

> If the block you click is at the deepest nesting level, clicking it again should exit back to edit mode. That is, if you click a Paragraph twice inside a Cover, you go back to editing it.

I agree with this, I'll see how feasible it is.

**Comment by youknowriad** [comment]
> If the block you click is at the deepest nesting level, clicking it again should exit back to edit mode. That is, if you click a Paragraph twice inside a Cover, you go back to editing it.

@jasmussen Turns out this is trickier than I originally thought because if the block has children blocks,  it doesn't mean all its content is made of inner blocks, one big part of it might be edited in the block itself (example Media & Text or Cover blocks) which make me think this is probably not the best approach.

I do think it's weird that you're able to edit text... in navigation mode though, so it makes me think, maybe the fix here is not to make the "edit mode" more easy to trigger but the other way around, making it harder to trigger without an explicit button somewhere in the UI (or by changing the tool).

So basically, when you're in a leaf node an the navigation tool is enabled,  you won't able to click the block to edit its content, unless you explicitely choose to go to the edit tool.

**Comment by jasmussen** [comment]
> I do think it's weird that you're able to edit text... in navigation mode though, so it makes me think, maybe the fix here is not to make the "edit mode" more easy to trigger but the other way around, making it harder to trigger without an explicit button somewhere in the UI (or by changing the tool).

The challenge here is that the Escape is intuitively something you use to _escape a mode_, and people are likely to want to escape the selection tool rather than escape the editing too. I don't think those shortcuts are easy to invert or change, but if it was, it seems like might be an option to make the selection tool have a different shortcut, and have Escape be the one that returns to the edit tool. 

However if we think creatively then perhaps we can find something more intuitive. 

What if you entered editing mode again by simply clicking the same block twice? One click selects it, another click edits it? This is Figma-esque, only they require double-clicks.

**Comment by youknowriad** [comment]
Pushed an update here, let me know how it feels @jasmussen

**Comment by jasmussen** [comment]
Thanks for the update. I'm not entirely sure what to look for here, it seems like I can now return to the edit mode ONLY using the Enter key, or selecting the tool in the toolbar. Is that intended? It feels slightly confusing.

**Comment by youknowriad** [comment]
I implemented this basically 

> What if you entered editing mode again by simply clicking the same block twice? One click selects it, another click edits it? This is Figma-esque, only they require double-clicks.

did you update your build properly? Checking again.

**Comment by jasmussen** [comment]
I got this when I looked more closely:

```
hint: 'try/select-edit-tool' matched more than one remote tracking branch.
hint: We found 2 remotes with a reference that matched. So we fell back
hint: on trying to resolve the argument as a path, but failed there too!
```

So yeah, problem on my end, looks like!

I got the branch to run, and I'm seeing this:

nesting

And I'm utterly loving it! It feels easy enough to get out of the mode, yet it feels solid enough to have a good "parent-first" block selection experience. This seems 👌  to me. 👏

With that in place, overall I personally feel like this is a massive improvement to Navigation mode. It keeps the accessibility benefits of navigation mode and expands them to mouse users. It adds a visual indicator to the toolbar, and I believe that we can further visually distinguish the two modes using some of the ideas found in #18667. 

One additional question, and not sure it's necessary for this branch: is there a way we can announce the mode switch to screen readers? 

For example when I enable VoiceOver (⌘F5 in MacOS, works best in Safari), I see the following message when the browser has focus:

Screenshot 2019-11-26 at 13 53 56

I would _love_ for us to be able to announce the mode switch to Navigation Mode in a similar fashion, something like:

> You are currently in navigation mode. Navigate blocks using arrow keys. To exit navigation mode and edit the selected block, press Enter. 

I will defer to experts on whether and how this would be possible and feasible, but it seems like we might be able to dynamically add an `alert` like this: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role

CC: @gziolo @enriquesanchez in case you have thoughts.

**Comment by enriquesanchez** [comment]
Agreed that mode switching needs to be communicated to AT. The `alert` role seems to be the way to go.

>You are currently in navigation mode. Navigate blocks using arrow keys. To exit navigation mode and edit the selected block, press Enter.

Love @jasmussen's copy suggestion. It's clear and descriptive.

Happy to help test this on both VoiceOver and NVDA.

**Comment by youknowriad** [comment]
I added a message for screen reader users using our existing a11y `speak` helper. Anything left here?

File: packages/block-editor/src/store/actions.js
Comment (gziolo) [inline packages/block-editor/src/store/actions.js]: Are you sure that it should be announced every time someone switches modes? When editing individual blocks with the keyboard it might be a common pattern to switch between modes.

Comment (gziolo) [reply packages/block-editor/src/store/actions.js]: > For example when I enable VoiceOver (⌘F5 in MacOS, works best in Safari), I see the following message when the browser has focus:
> 
> Screenshot 2019-11-26 at 13 53 56

I see it now. We can give it a try. Someone should battle test it though :)

**Comment by gziolo** [comment]
> there is also one usability challenge, the shortcut for Select mode is esc, however when focus is in the dropdown menu – esc will close the menu, it might be confusing as some user might assume that this will change the mode, I don't have a solution for it

This is still an issue. Actually, it's wrong for both tools as they use very popular keyboard shortcuts `esc` and `enter`. The question is whether we should show those keyboard shortcuts in the menu knowing that folks can't use them in this context as advertised?

File: packages/block-editor/src/store/actions.js
Comment (jasmussen) [reply packages/block-editor/src/store/actions.js]: It is my understanding that yes, we do need something like this. We received helpful information from a user in https://github.com/WordPress/gutenberg/issues/17088#issuecomment-557873792, who had also created #18583 which was related to the switch.

**Comment by youknowriad** [comment]
> The question is whether we should show those keyboard shortcuts in the menu knowing that folks can't use them in this context as advertised?

That's a good question, we could potentially add new global shortcuts or remove the shortcuts from there now. What do you think @jasmussen

**Comment by jasmussen** [comment]
> The question is whether we should show those keyboard shortcuts in the menu knowing that folks can't use them in this context as advertised?

Changing one or both of the Esc and Enter shortcuts is an option. Adding additional shortcuts is another. I would suggest that this could/should be explored and discussed separately, as this very pull request addressed an issue already shipping in master, and that it's worth getting that benefit in soon. 

> This is still an issue. Actually, it's wrong for both tools as they use very popular keyboard shortcuts esc and enter. The question is whether we should show those keyboard shortcuts in the menu knowing that folks can't use them in this context as advertised?

I'm unsure how to best address this, and even _whether_ we should address it. It is a question of local vs. global shortcuts. Take ⌘B for "bold", for example, this is a local shortcut in that it only works when the cursor is in text. This is similar to the current navigatin mode behavior, as it requires a block to be selected first. And then there is the duality of Escape _also_ meaning "close menu" or "clear selection" (when you've selected text).

So should we show the shortcut hint if it doesn't work when you're viewing the shortcut hint? I don't know... in the Keyboard Shortcuts sheet you can see the shortcut hint for bold even if it doesn't work there. But at least that sheet has clear context such as "Global Shortcuts", etc.

Question: can we make escape into a global shortcut? Can we make it _both invoke navigation mode_ AND close the menu? If we can, should we?

Alternately, let's try this:

- Add the new shortcuts to the Keyboad Shortcuts dialog, under "block shortcuts" or "selection shortcuts".
- I think we should keep the menu hint, but revisit if it remains a problem.
- If Esc and Enter remain challenging due to their existing actions, we should revisit those separately.

Sound good?

**Comment by gziolo** [comment]
This one is the most interesting when focus is on the selected "Select" mode: 
Screen Shot 2019-11-27 at 10 57 20

Pressing `Enter` will only work as expected when you move the focus to the "Edit" item.

When switching to "Edit" mode with the keyboard, you rather don't want `Esc` to get activated to switch back to "Select" mode but rather the intention is to close the menu:
modes

I would personally remove those shortcuts from this menu and put them in the modal dialog which presents all keyboard shortcuts.

**Comment by jasmussen** [comment]
> I would personally remove those shortcuts from this menu and put them in the modal dialog which presents all keyboard shortcuts.

Another option emerges. Instead of writing:

> Tools offer different block interactions to optimize block selection & editing tasks.

... we could remove the shortcuts hints from the right side, and instead write the following in the prose:

```
Tools offer different interactions to optimize block selection & editing tasks.

When editing, press <key>Escape</key> to select it with the Selection tool. Press <key>Enter</key> to go back to editing.
```

We'd want it _also_ in the block keyboard shortcuts help sheet, but what do you think about the above?

File: packages/block-editor/src/store/actions.js
Comment (gziolo) [reply packages/block-editor/src/store/actions.js]: Right, my concern is mostly about the number of details included in the message. Anyway, let's collect feedback and refine only if necessary.

**Comment by youknowriad** [comment]
it doesn't feel great

Capture d’écran 2019-11-27 à 11 56 58 AM

**Comment by jasmussen** [comment]
Yeah that's a lot. I think it's still a good place. Maybe we can try this:

> Tools offer different interactions for block selection & editing. To select, press Escape, to go back to editing, press Enter. 

What do you think?

**Comment by youknowriad** [comment]
Pushed it, feel free to tweak it as needed.

**Comment by jasmussen** [comment]
I think it's good to try!

**Comment by youknowriad** [comment]
Approve :P

**Review by jasmussen — APPROVED** [review]
This PR in my opinion provides a solid accessibility improvement to navigation mode shipping in master. It provides a visual indicator, and better clarification on how to switch between the two. It also does not preclude us from making further improvements in the future!

Before you merge, it would be good to have @enriquesanchez just run through the announcement that happens to screenreaders when you switch modes. Other than that, 👍 👍 and thank you.

**Review by gziolo — APPROVED** [review]
Nice work @youknowriad. It's a huge challenge to integrate those tools in the UI. I'm not surprised it triggered a lot of discussions. 😃 

Code wise, everything looks good and all my comments were addressed 👍 

Let's gather the initial feedback and refine the UI bits separately.

**Comment by noisysocks** [comment]
It's confusing to me that when I'm in selection mode I still see a text cursor (`cursor: text`) while hovering over paragraph blocks. Could we set `cursor: default` on non-selected blocks when in selection mode?

**Comment by youknowriad** [comment]
@noisysocks That's a good point 👍

**Comment by enriquesanchez** [comment]
Hey @youknowriad! 

I'm just catching up with my notifications and looks like this branch was already merged and deleted. I'm not sure if there's still a way for me to test this, let me know.

**Comment by youknowriad** [comment]
@enriquesanchez It's in master, so you can try master directly and we can follow-up with any problems you find.

**Comment by ZebulanStanphill** [comment]
When hovering a nested block in select mode, the left border disappears:

image

**Comment by youknowriad** [comment]
@ZebulanStanphill this seems like a small design glitch, would you mind creating an issue?

**Comment by ZebulanStanphill** [comment]
@youknowriad Done: #18824

File: packages/e2e-test-utils/src/keyboard-mode.js
Comment (aduth) [inline packages/e2e-test-utils/src/keyboard-mode.js]: Should this file have been removed? It remains as an empty, unreferenced file.

Comment (youknowriad) [reply packages/e2e-test-utils/src/keyboard-mode.js]: 😅

**Pull request merged**
